### PR TITLE
[Backport v2.7-branch] Bluetooth: fix HCI command / data deadlock

### LINKED ISSUE #57947: Backport Bluetooth fixes to 2.7 branch
**Is your feature request related to a problem? Please describe.**
I am working on a project that uses the Zephyr 2.7 release, but have run into several known issues that have been fixed in later releases but are showstoppers for me on 2.7. I would like fixes for these issues to be applied to the 2.7 branch.

**Describe the solution you'd like**
Backport the following PRs to 2.7:
#52536
#48395
#46165

I see there has been at least one other request for some of these PRs to be backported so I think there are others in the community that would benefit:
https://github.com/zephyrproject-rtos/zephyr/issues/43718#issuecomment-1470331996

**Describe alternatives you've considered**
I have attempted backporting these changes myself, but I don't have expertise in the Bluetooth stack and am not confident that my changes are completely effective nor that I have included everything necessary to fix the issues (since there were other changes between the 2.7 release and these PRs being merged that cause conflicts). I could potentially submit my attempt at backporting these changes so they could be reviewed and used by the community, but would need advice on the best way to do that.

Using a more recent release (or main) is not an option for me at this time.

**Additional context**
None

### LINKED ISSUE #58412: [Backport v2.7-branch] Failed to backport #52536
This issue was created automatically because the backport of #52536 to `v2.7-branch` failed.

The backport to `v2.7-branch` failed:
```
Auto-merging subsys/bluetooth/host/conn.c
CONFLICT (content): Merge conflict in subsys/bluetooth/host/conn.c
error: could not apply c3e5fabbf1... Bluetooth: host: make HCI fragmentation async
hint: After resolving the conflicts, mark them with
hint: "git add/rm <pathspec>", then run
hint: "git cherry-pick --continue".
hint: You can instead skip this commit with "git cherry-pick --skip".
hint: To abort and get back to the state before "git cherry-pick",
hint: run "git cherry-pick --abort".
```
To backport manually, run these commands in your terminal:
```bash
# Fetch latest updates from GitHub
git fetch
# Create a new working tree
git worktree add .worktrees/backport-v2.7-branch v2.7-branch
# Navigate to the new working tree
cd .worktrees/backport-v2.7-branch
# Create a new branch
git switch --create backport-52536-to-v2.7-branch
# Cherry-pick the merged commits of this pull request and resolve the conflicts
git cherry-pick -x 8bc094610c5dd69f6ba0085fb3fe7d18167164c9~4..8bc094610c5dd69f6ba0085fb3fe7d18167164c9
# Push it to GitHub
git push --set-upstream origin backport-52536-to-v2.7-branch
# Go back to the original working tree
cd ../..
# Delete the working tree
git worktree remove .worktrees/backport-v2.7-branch
```
Then, create a pull request where the `base` branch is `v2.7-branch` and the `compare`/`head` branch is `backport-52536-to-v2.7-branch`.

### PR DESCRIPTION
**Pull request by theob-pro**
Fixes https://github.com/zephyrproject-rtos/zephyr/issues/58412.

The following commits from https://github.com/zephyrproject-rtos/zephyr/pull/50476 have been added:
- https://github.com/zephyrproject-rtos/zephyr/commit/7a6872d83729360d2061957305e68250107c9e64
- https://github.com/zephyrproject-rtos/zephyr/commit/1c8fe67a52766381fa02de4a39b1888c18d63b6a
- https://github.com/zephyrproject-rtos/zephyr/commit/3c1ca93fe8342d004bf539e3c193efb034ba448a
- https://github.com/zephyrproject-rtos/zephyr/commit/8e207fefad2e1aeaaf03ee4da0fb024d167b7a96
- https://github.com/zephyrproject-rtos/zephyr/commit/77e1a9dcadff4555c5fd9d5ecef10cd7d9fb1c28

https://github.com/zephyrproject-rtos/zephyr/pull/52760 is also backported with this PR.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/57947

### CODE REVIEW

File: subsys/bluetooth/host/conn.c [hunk 1]
```diff
@@ -520,46 +543,61 @@ static bool send_frag(struct bt_conn *conn, struct net_buf *buf, uint8_t flags,
...
+
+		net_buf_add_mem(frag, buf->data, frag_len);
+		net_buf_pull(buf, frag_len);
+	} else {
+		/* De-queue the buffer now that we know we can send it.
+		 * Only applies if the buffer to be sent is the original buffer,
+		 * and not one of its fragments.
+		 * This buffer was fetched from the FIFO using a peek operation.
+		 */
+		buf = net_buf_get(&conn->tx_queue, K_NO_WAIT);
```
Comment (jori-nordic) [inline subsys/bluetooth/host/conn.c:591]: If I remember, this relied on the patch that removed the `frags` union in `struct net_buf` (https://github.com/zephyrproject-rtos/zephyr/pull/52760).
Maybe backport that too?

**Review by jori-nordic — CHANGES_REQUESTED** [review]
see comment

**Review by jori-nordic — APPROVED** [review]

**Comment by cfriedt** [comment]
@jhedberg - looks to be ready to me, but this will need your approval. Please take a look.

**Review by jhedberg — APPROVED** [review]

**Comment by sheindel-llt** [comment]
Thanks for doing this packport!

**Pull request merged**